### PR TITLE
Normal and Uniform return earlier without entering kernel<RNG>

### LINKED ISSUE #103418: Normal return earlier without entering normal_kernel<RNG>
### 🚀 The feature, motivation and pitch

Hello, I noticed that when we call the normal function with an empty shape, it produces an empty tensor. For example:

```py
import torch

a = torch.randn(5, 0)
print(a)            # tensor([], size=(5, 0))
print(a.numel())    # 0
print(hex(a.untyped_storage().data_ptr()))  # 0x0
```

In this case, calling the normal_kernel seems unnecessary as it consumes time and resources. It appears that the normal_kernel invokes the `tensor iterator` and `serial_for_each`, which could be relatively costly.

Would it be beneficial to add some code to the `normal_impl_` function in `aten/src/ATen/native/DistributionTemplates.h` to return earlier in such cases?

### Alternatives

```c++
template<template<typename> class normal_kernel, typename RNG>
Tensor& normal_impl_(Tensor& self, double mean, double std, c10::optional<Generator> gen) {
  CHECK_NORMAL_STD(std);
  /* ####code added here#### */
  if (self.numel() == 0) {
    return self;
  }
  if (self.is_complex()) {
    auto float_tensor = at::view_as_real(self);
    // variance for normal distribution of the real and imaginary values
    // is half of the input variance
    normal_kernel<RNG>()(float_tensor, mean, std/(std::sqrt(2)), gen);
  } else {
    normal_kernel<RNG>()(self, mean, std, gen);
  }
  return self;
}
```

### Additional context

企业微信截图_f7087827-50d0-42e7-9d23-fce9a31e755c

cc @fritzo @neerajprad @alicanb @nikitaved

### PR DESCRIPTION
**Pull request by yewentao256**
Fixes [#103418](https://github.com/pytorch/pytorch/issues/103418)

By the way, I'm wondering should we update other distribution funcs in `aten\src\ATen\native\DistributionTemplates.h`?

### CODE REVIEW

**Review by mikaylagawarecki — COMMENTED** [review]
Should this apply only to Normal and Uniform or do we want to enforce this across other impls in this file (e.g. exponential etc.) Seems like they mostly also invoke TensorIterator so should run into the same issue mentioned by https://github.com/pytorch/pytorch/issues/103418

**Comment by yewentao256** [comment]
> Should this apply only to Normal and Uniform or do we want to enforce this across other impls in this file (e.g. exponential etc.) Seems like they mostly also invoke TensorIterator so should run into the same issue mentioned by #103418

Sure, I'll apply to all of them in `aten\src\ATen\native\DistributionTemplates.h`

How about making a micro to check and return earlier?

**Review by jgong5 — CHANGES_REQUESTED** [review]
Please take care of other functions as @mikaylagawarecki suggested.

> How about making a micro to check and return earlier?

Either way is fine. There are not many functions anyway.

**Comment by yewentao256** [comment]
> Either way is fine. There are not many functions anyway.

Done~

File: aten/src/ATen/native/DistributionTemplates.h [hunk 1]
```diff
@@ -184,6 +191,11 @@ at::Tensor& random_from_to_impl(at::Tensor& self, int64_t from, c10::optional<in
 template<template<typename> class normal_kernel, typename RNG>
 Tensor& normal_impl_(Tensor& self, double mean, double std, c10::optional<Generator> gen) {
   CHECK_NORMAL_STD(std);
+
+  // return earlier for not invoking normal_kernel
+  // see https://github.com/pytorch/pytorch/issues/103418 for more details
```
Comment (jgong5) [inline aten/src/ATen/native/DistributionTemplates.h:196]: A bit weird to only leave the comment in `normal_impl_` while this logic applies to other functions here too. Maybe revise and move the comment to the definition of `CHECK_EMPTY_AND_RETURN`?

**Review by jgong5 — APPROVED** [review]

**Comment by yewentao256** [comment]
Hello jgong5 @jgong5, I am wondering is there anything else I need to do? Otherwise could this pr be merged? Thanks!

**Review by mikaylagawarecki — COMMENTED** [review]
The failures on HUD seem related https://hud.pytorch.org/pr/103507 

For example, [this assertion](https://github.com/pytorch/pytorch/blob/main/test/test_tensor_creation_ops.py#L2707) is failing. Perhaps an empty shape with the wrong dtype should still be raising this RuntimeError but the new macro has bypassed this check. Could you update accordingly to fix this please 

Also cc @jgong5 for thoughts

**Comment by yewentao256** [comment]
@mikaylagawarecki Thank you for pointing out this issue. I have addressed it by moving the empty return statement after the original check (dtype, etc.). Please let me know if there is anything else that needs to be fixed.

**Review by mikaylagawarecki — APPROVED** [review]

**Pull request closed without merging**